core/src/Kokkos_LogicalMemorySpace.hpp

```diff
@@ -0,0 +1,358 @@
+
+#include <cstring>
+#include <string>
+#include <iosfwd>
+#include <typeinfo>
+
+#include <Kokkos_Core_fwd.hpp>
+#include <Kokkos_Concepts.hpp>
+#include <Kokkos_MemoryTraits.hpp>
+
+#include <impl/Kokkos_Traits.hpp>
+#include <impl/Kokkos_Error.hpp>
+#include <impl/Kokkos_SharedAlloc.hpp>
+
+#include "impl/Kokkos_HostSpace_deepcopy.hpp"
+
+/*--------------------------------------------------------------------------*/
+
+namespace Kokkos {
+
+// TODO DZP fix docs
+/// \class LogicalMemorySpace
+/// \brief
+///
+/// LogicalMemorySpace is
+
+template <const char* const* Name, class BaseSpace,
+          class DefaultExecutionSpace = void, bool SharesAccessWithBase = true>
+class LogicalMemorySpace {
+ public:
+  //! Tag this class as a kokkos memory space
+  typedef LogicalMemorySpace<Name, BaseSpace, DefaultExecutionSpace,
+                             SharesAccessWithBase>
+      memory_space;
+  typedef size_t size_type;
+
+  /// \typedef execution_space
+  /// \brief Default execution space for this memory space.
+  ///
+  /// Every memory space has a default execution space.  This is
+  /// useful for things like initializing a View (which happens in
+  /// parallel using the View's default execution space).
+
+  using execution_space =
+      typename std::conditional<std::is_void<DefaultExecutionSpace>::value,
+                                typename BaseSpace::execution_space,
+                                DefaultExecutionSpace>::type;
+
+  typedef Kokkos::Device<execution_space, memory_space> device_type;
+
+  /**\brief  Default memory space instance */
+  LogicalMemorySpace() : underlying_allocator(){};
+  LogicalMemorySpace(LogicalMemorySpace&& rhs)      = default;
+  LogicalMemorySpace(const LogicalMemorySpace& rhs) = default;
+  LogicalMemorySpace& operator=(LogicalMemorySpace&&) = default;
+  LogicalMemorySpace& operator=(const LogicalMemorySpace&) = default;
+  ~LogicalMemorySpace()                                    = default;
+
+  BaseSpace underlying_allocator;
+
+  template <typename... Args>
+  LogicalMemorySpace(Args&&... args) : underlying_allocator(args...){};
+
+  /**\brief  Allocate untracked memory in the space */
+  void* allocate(const size_t arg_alloc_size) const {
+    return underlying_allocator.allocate(arg_alloc_size);
+  }
+
+  /**\brief  Deallocate untracked memory in the space */
+  void deallocate(void* const arg_alloc_ptr,
+                  const size_t arg_alloc_size) const {
+    return underlying_allocator.deallocate(arg_alloc_ptr, arg_alloc_size);
+  }
+
+  /**\brief Return Name of the MemorySpace */
+  constexpr const char* name() const { return m_name; }
+
+ private:
+  constexpr static const char* m_name = *Name;
+  friend class Kokkos::Impl::SharedAllocationRecord<memory_space, void>;
+};
+
+}  // namespace Kokkos
+
+//----------------------------------------------------------------------------
+
+namespace Kokkos {
+
+namespace Impl {
+
+template <const char* const* Name, typename BaseSpace,
```

Comment:
@dhollman: I'm going to start looking at merging this, but I still need help with this part. Down on Line 149, I call out HostSpace specifically, but that's wrong. It needs to just be `OtherSpace` like in this specialization. The problem is that if I do that, I get ambiguous specializations in the case where `OtherSpace = Kokkos::LogicalMemorySpace...`

I couldn't find a way to make it so that only one of these would work in such a case without modifying the base MemorySpaceAccess template. Is there a way to make it work, or should I just exercise some light chainsaw surgery on the base template? I've got a godbolt representing the problem here

https://godbolt.org/z/DSYu5g

SFINAE-fu gurudom would be appreciated

Reply:
I like your Godbolt. As it literally can't mess up anything else, while I just can't think of a way for the default template parameter to mess stuff up, I'll use that. I'm a little surprised that works, is it a "this is more specialized" thing? I really need to read up on those rules, they're interesting

Reply:
Yes, this is a "this is more specialized" thing, and no, you probably don't need to read up on the rules (unless you're going to implement them in a compiler)—they should just generally work as you expect.  There are a few surprises here and there, but if something "looks" more specific, it probably is.